commit f67db227af31bba7cdf2a7a00b97af91b588c2f5
Author: Zoltan Herczeg <zherczeg7@gmail.com>
Date:   Sun Aug 9 11:05:54 2026 +0200

    Fix pcre2_match to check for JIT support before JIT validation & execution (#926)
    
    This fixes the issue that the JIT branch's UTF validation is not pinned to be identical to the interpreter's validation.
    
    This was not robust, and lead to a bug, in the case where the JIT UTF validation is done, but because the relevant JIT mode was not compiled, it falls through to the interpreter and skips the interpreter's own UTF validation and setup.

diff --git a/src/pcre2_internal.h b/src/pcre2_internal.h
index 2e8c7e47..930c745b 100644
--- a/src/pcre2_internal.h
+++ b/src/pcre2_internal.h
@@ -2301,6 +2301,7 @@ is available. */
 #define _pcre2_is_newline            PCRE2_SUFFIX(_pcre2_is_newline_)
 #define _pcre2_jit_free_rodata       PCRE2_SUFFIX(_pcre2_jit_free_rodata_)
 #define _pcre2_jit_free              PCRE2_SUFFIX(_pcre2_jit_free_)
+#define _pcre2_jit_check_exec        PCRE2_SUFFIX(_pcre2_jit_check_exec_)
 #define _pcre2_jit_get_size          PCRE2_SUFFIX(_pcre2_jit_get_size_)
 #define _pcre2_jit_get_target        PCRE2_SUFFIX(_pcre2_jit_get_target_)
 #define _pcre2_memctl_malloc         PCRE2_SUFFIX(_pcre2_memctl_malloc_)
@@ -2330,6 +2331,7 @@ extern BOOL         _pcre2_is_newline(PCRE2_SPTR, uint32_t, PCRE2_SPTR,
                       uint32_t *, BOOL);
 extern void         _pcre2_jit_free_rodata(void *, void *);
 extern void         _pcre2_jit_free(void *, pcre2_memctl *);
+extern BOOL         _pcre2_jit_check_exec(void *, uint32_t);
 extern size_t       _pcre2_jit_get_size(void *);
 const char *        _pcre2_jit_get_target(void);
 extern void *       _pcre2_memctl_malloc(size_t, pcre2_memctl *);
diff --git a/src/pcre2_jit_match_inc.h b/src/pcre2_jit_match_inc.h
index 4163cf61..ba210007 100644
--- a/src/pcre2_jit_match_inc.h
+++ b/src/pcre2_jit_match_inc.h
@@ -117,6 +117,7 @@ jit_arguments arguments;
 int rc;
 int index = 0;
 
+/* The same check is performed by jit_check_exec(). */
 if ((options & PCRE2_PARTIAL_HARD) != 0)
   index = 2;
 else if ((options & PCRE2_PARTIAL_SOFT) != 0)
diff --git a/src/pcre2_jit_misc_inc.h b/src/pcre2_jit_misc_inc.h
index 0225fc6b..16c230e9 100644
--- a/src/pcre2_jit_misc_inc.h
+++ b/src/pcre2_jit_misc_inc.h
@@ -200,17 +200,28 @@ if (jit_stack != NULL)
 
 
 /*************************************************
-*               Get target CPU type              *
+*          Checks function compilation           *
 *************************************************/
 
-const char*
-PRIV(jit_get_target)(void)
+BOOL
+PRIV(jit_check_exec)(void *executable_jit, uint32_t options)
 {
 #ifndef SUPPORT_JIT
-return "JIT is not supported";
+(void)executable_jit;
+(void)options;
+return FALSE;
 #else  /* SUPPORT_JIT */
-return sljit_get_platform_name();
-#endif  /* SUPPORT_JIT */
+/* The same check is performed at the beginning of pcre2_jit_match(). */
+executable_functions *functions = (executable_functions *)executable_jit;
+int index = 0;
+
+if ((options & PCRE2_PARTIAL_HARD) != 0)
+  index = 2;
+else if ((options & PCRE2_PARTIAL_SOFT) != 0)
+  index = 1;
+
+return functions->executable_funcs[index] != NULL;
+#endif
 }
 
 
@@ -231,4 +242,19 @@ return executable_sizes[0] + executable_sizes[1] + executable_sizes[2];
 #endif
 }
 
+/*************************************************
+*               Get target CPU type              *
+*************************************************/
+
+const char*
+PRIV(jit_get_target)(void)
+{
+#ifndef SUPPORT_JIT
+return "JIT is not supported";
+#else  /* SUPPORT_JIT */
+return sljit_get_platform_name();
+#endif  /* SUPPORT_JIT */
+}
+
+
 /* End of pcre2_jit_misc_inc.h */
diff --git a/src/pcre2_match.c b/src/pcre2_match.c
index 9ee8a476..a5a8421f 100644
--- a/src/pcre2_match.c
+++ b/src/pcre2_match.c
@@ -7005,10 +7005,6 @@ PCRE2_SPTR req_cu_ptr;
 PCRE2_SPTR start_partial;
 PCRE2_SPTR match_partial;
 
-#ifdef SUPPORT_JIT
-BOOL use_jit;
-#endif
-
 /* This flag is needed even when Unicode is not supported for convenience
 (it is used by the IS_NEWLINE macro). */
 
@@ -7018,9 +7014,6 @@ BOOL utf = FALSE;
 BOOL ucp = FALSE;
 BOOL allow_invalid;
 uint32_t fragment_options = 0;
-#ifdef SUPPORT_JIT
-BOOL jit_checked_utf = FALSE;
-#endif
 #endif  /* SUPPORT_UNICODE */
 
 PCRE2_SIZE frame_size;
@@ -7083,15 +7076,6 @@ options |= (re->flags & FF) / ((FF & (~FF+1)) / (OO & (~OO+1)));
 #undef FF
 #undef OO
 
-/* If the pattern was successfully studied with JIT support, we will run the
-JIT executable instead of the rest of this function. Most options must be set
-at compile time for the JIT code to be usable. */
-
-#ifdef SUPPORT_JIT
-use_jit = (re->executable_jit != NULL &&
-          (options & ~PUBLIC_JIT_MATCH_OPTIONS) == 0);
-#endif
-
 /* Initialize UTF/UCP parameters. */
 
 #ifdef SUPPORT_UNICODE
@@ -7138,20 +7122,25 @@ match_data->startchar = 0;
 
 /* ============================= JIT matching ============================== */
 
-/* Prepare for JIT matching. Check a UTF string for validity unless no check is
-requested or invalid UTF can be handled. We check only the portion of the
-subject that might be be inspected during matching - from the offset minus the
-maximum lookbehind to the given length. This saves time when a small part of a
-large subject is being matched by the use of a starting offset. Note that the
-maximum lookbehind is a number of characters, not code units. */
+/* If the pattern was successfully studied with JIT support, we will run the
+JIT executable instead of the rest of this function. Most options must be set
+at compile time for the JIT code to be usable. */
 
 #ifdef SUPPORT_JIT
-if (use_jit)
+if (re->executable_jit != NULL &&
+    (options & ~PUBLIC_JIT_MATCH_OPTIONS) == 0 &&
+    PRIV(jit_check_exec)(re->executable_jit, options))
   {
+  /* Prepare for JIT matching. Check a UTF string for validity unless no check
+  is requested or invalid UTF can be handled. We check only the portion of the
+  subject that might be be inspected during matching - from the offset minus
+  the maximum lookbehind to the given length. This saves time when a small part
+  of a large subject is being matched by the use of a starting offset. Note that
+  the maximum lookbehind is a number of characters, not code units. */
+
 #ifdef SUPPORT_UNICODE
   if (utf && (options & PCRE2_NO_UTF_CHECK) == 0 && !allow_invalid)
     {
-
     /* For 8-bit and 16-bit UTF, check that the first code unit is a valid
     character start. */
 
@@ -7204,40 +7193,36 @@ if (use_jit)
       match_data->startchar += start_match - subject;
       return match_data->rc = rc;
       }
-    jit_checked_utf = TRUE;
     }
 #endif  /* SUPPORT_UNICODE */
 
-  /* If JIT returns BADOPTION, which means that the selected complete or
-  partial matching mode was not compiled, fall through to the interpreter. */
-
   rc = pcre2_jit_match(code, subject, length, start_offset, options,
     match_data, mcontext);
-  if (rc != PCRE2_ERROR_JIT_BADOPTION)
+  /* JIT must be able to perform the match. */
+  PCRE2_ASSERT(rc != PCRE2_ERROR_JIT_BADOPTION);
+
+  match_data->options = original_options;
+  if (rc >= 0 && (options & PCRE2_COPY_MATCHED_SUBJECT) != 0)
     {
-    match_data->options = original_options;
-    if (rc >= 0 && (options & PCRE2_COPY_MATCHED_SUBJECT) != 0)
+    if (length != 0)
       {
-      if (length != 0)
-        {
-        match_data->subject = match_data->memctl.malloc(CU2BYTES(length),
-          match_data->memctl.memory_data);
-        if (match_data->subject == NULL)
-          return match_data->rc = PCRE2_ERROR_NOMEMORY;
-        memcpy((void *)match_data->subject, subject, CU2BYTES(length));
-        }
-      else
-        match_data->subject = NULL;
-      match_data->flags |= PCRE2_MD_COPIED_SUBJECT;
+      match_data->subject = match_data->memctl.malloc(CU2BYTES(length),
+        match_data->memctl.memory_data);
+      if (match_data->subject == NULL)
+        return match_data->rc = PCRE2_ERROR_NOMEMORY;
+      memcpy((void *)match_data->subject, subject, CU2BYTES(length));
       }
     else
-      {
-      /* When pcre2_jit_match sets the subject, it doesn't know what the
-      original passed-in pointer was. */
-      if (match_data->subject != NULL) match_data->subject = original_subject;
-      }
-    return rc;
+      match_data->subject = NULL;
+    match_data->flags |= PCRE2_MD_COPIED_SUBJECT;
     }
+  else
+    {
+    /* When pcre2_jit_match sets the subject, it doesn't know what the
+    original passed-in pointer was. */
+    if (match_data->subject != NULL) match_data->subject = original_subject;
+    }
+  return rc;
   }
 #endif  /* SUPPORT_JIT */
 
@@ -7250,12 +7235,8 @@ this. */
 
 mb->check_subject = subject;
 
-/* If a UTF subject string was not checked for validity in the JIT code above,
-check it here, and handle support for invalid UTF strings. The check above
-happens only when invalid UTF is not supported and PCRE2_NO_CHECK_UTF is unset.
-If we get here in those circumstances, it means the subject string is valid,
-but for some reason JIT matching was not successful. There is no need to check
-the subject again.
+/* Check the validity of UTF subject strings. The check happens only when
+PCRE2_NO_CHECK_UTF is unset.
 
 We check only the portion of the subject that might be be inspected during
 matching - from the offset minus the maximum lookbehind to the given length.
@@ -7267,11 +7248,7 @@ Note also that support for invalid UTF forces a check, overriding the setting
 of PCRE2_NO_CHECK_UTF. */
 
 #ifdef SUPPORT_UNICODE
-if (utf &&
-#ifdef SUPPORT_JIT
-    !jit_checked_utf &&
-#endif
-    ((options & PCRE2_NO_UTF_CHECK) == 0 || allow_invalid))
+if (utf && ((options & PCRE2_NO_UTF_CHECK) == 0 || allow_invalid))
   {
 #if PCRE2_CODE_UNIT_WIDTH != 32
   BOOL skipped_bad_start = FALSE;

commit e2f100156b94cf85dce9046b47a3274590487de4
Author: Nicholas Wilson <nicholas@nicholaswilson.me.uk>
Date:   Wed Aug 12 10:18:54 2026 +0100

    Add a testcase for #926 (#940)

diff --git a/RunTest b/RunTest
index 57a03455..306427b6 100755
--- a/RunTest
+++ b/RunTest
@@ -175,9 +175,12 @@ checkresult()
     *)    with="";;
   esac
   cf_out="$testdata/testoutput$2"
-  if [ $ebcdic -eq 1 ] ; then
-    # We currently only use the #if ... #endif support in pcre2test for EBCDIC
-    # testing. Run in "preprocess-only" mode (-E) on the testoutput file to trim
+  if [ $ebcdic -eq 1 ] || [ $utf -eq 0 ]; then
+    # We currently only use the #if ... #endif support in pcre2test for tests
+    # which are excluded in certain builds:
+    #   - EBCDIC
+    #   - No-UTF
+    # Run in "preprocess-only" mode (-E) on the testoutput file to trim
     # the output lines matching the input lines which are discarded.
     $sim $pcre2test -q -E "$cf_out" >testoutput$bits$3/testoutput$2-expected
     cf_out=testoutput$bits$3/testoutput$2-expected
diff --git a/src/pcre2test_inc.h b/src/pcre2test_inc.h
index a74e3368..72767916 100644
--- a/src/pcre2test_inc.h
+++ b/src/pcre2test_inc.h
@@ -1712,7 +1712,7 @@ process_command(void)
 FILE *f;
 PCRE2_SIZE serial_size;
 size_t i;
-int rc, cmd, yield;
+int rc, cmd, yield, config_value;
 uint16_t first_listed_newline;
 const char *cmdname;
 size_t cmdlen;
@@ -1980,7 +1980,7 @@ switch(cmd)
     {
     size_t optlen = strlen(coptlist[i].name);
     const uint8_t *argptr_trail;
-    if (coptlist[i].type != CONF_FIX)
+    if (coptlist[i].type != CONF_FIX && coptlist[i].type != CONF_INT)
       continue;
     if (strncmp((const char*)argptr, coptlist[i].name, optlen) != 0)
       continue;
@@ -1995,9 +1995,20 @@ switch(cmd)
     return PR_ABEND;
     }
 
-  /* Condition FALSE - skip this line and everything until #endif. */
-  if ((coptlist[i].value != 0) == if_inverted)
+  if (coptlist[i].type == CONF_FIX)
+    {
+    config_value = coptlist[i].value;
+    }
+  else /* if (coptlist[i].type == CONF_INT) */
+    {
+    (void)pcre2_config(coptlist[i].value, &config_value);
+    }
+
+  if ((config_value != 0) == if_inverted)
+    {
+    /* Condition FALSE - skip this line and everything until #endif. */
     yield = PR_ENDIF;
+    }
 
   inside_if = TRUE;
   break;
diff --git a/testdata/testinput17 b/testdata/testinput17
index 9d728965..e6dbbe42 100644
--- a/testdata/testinput17
+++ b/testdata/testinput17
@@ -336,4 +336,14 @@
     zabcz\=callout_error=1
     zabcz\=callout_error=1,substitute_matched
 
+# Check UTF validation when falling back from a partially-compiled JIT pattern
+# to the interpreter
+
+#if unicode
+
+/\b/B,utf,ucp,jit=1
+    \xff\x00\x00\=offset=2,ps
+
+#endif
+
 # End of testinput17
diff --git a/testdata/testoutput17 b/testdata/testoutput17
index 773ec18a..11f032d8 100644
--- a/testdata/testoutput17
+++ b/testdata/testoutput17
@@ -603,4 +603,21 @@ Failed: error -37: callout error code
   1  ^        a
 Failed: error -37: callout error code
 
+# Check UTF validation when falling back from a partially-compiled JIT pattern
+# to the interpreter
+
+#if unicode
+
+/\b/B,utf,ucp,jit=1
+------------------------------------------------------------------
+        Bra
+        \b (ucp)
+        Ket
+        End
+------------------------------------------------------------------
+    \xff\x00\x00\=offset=2,ps
+Partial match: 
+
+#endif
+
 # End of testinput17
